chore: add pre releases config and job (#2874)
contentful/forma-36: RELEASES.md

Trying to make prereleases easier to handle we created a script that you can use on your branch before merging into master, that will bump the package you select, and you can choose if you it's an alpha or beta release, before publishing it to NPM.

You can check the script [here](https://github.com/contentful/forma-36/blob/c6b10071959a085b21e49f5411a5ebff2f8a70d6/scripts/prerelease.mjs)

## Next branch

The next branch is used for the new version of Forma 36, which will probably have breaking changes. To help the development of the new packages, we have setup an automatic publishing of the packages on that branch to the `alpha` tag.
These packages aren't shown on the website as they can have breaking changes often.

### How to automatic publish

To make changeset publish a new version of the packages it's the same as in the `main` branch, it just requires having a changeset generated and specifing the packages.

```sh
npx changeset
# Always select major change
```

Differently from the `main` branch, neither the packages changelog or the **What's new page** will be updated, and the changeset files will not be removed.

### Testing how the version would change locally

> It's possible to test how the version would change locally, make sure that your last changes are commited, and you can run the following commands depending on what needs to be tested.
> 
> Remember to revert the changes made through the `npx changeset version` and not commit them.

Reviewer: ```suggestion
After committing your work, test the version's change by running these commands.

> [!WARNING]
> Remember to revert the changes made through the `npx changeset version` and not commit them.
```

```sh
npx changeset;
# Select the package and set as major version
# Commit the changeset
git add .changeset; git commit -m 'chore: add changeset for [package name]'

npx changeset version;
# This will update the pre.json and the package.json file for the selected package.
# If the package is already in alpha, it will update the version number after -alpha, e.g. alpha.0 > alpha.1
# If the package wasn't in alpha yet, the version will be updated to the next major version and -alpha.0 added at the end.
```

#### Moving to stable version

After having some pre release changsets created like on the previous step, you can run the following to look how it would look like when moving to stable.

```sh
npx changeset pre exit;
# This updates the pre.json file to mark it as exiting prereleases mode

# To test how it would update the changelogs and how it would behave as in main branch, replace the .changeset/config.json file with the one from the main branch, before running changeset version

npx changeset version;
# This will remove the pre.json file, and update all package.json files that need to be updated to the version without the -alpha tag.

```

### Changeset configuration

> To make it possible we take advantage of changeset feature of prereleasing, to make it possible we have the [.changeset/pre.json](./.changeset/pre.json) which specifies which tag to use and the initial version of the packages, which is used to calculate the alpha version, for example:

Reviewer: ```suggestion
We use the pre-releasing changeset feature and the [.changeset/pre.json](./.changeset/pre.json), which specifies which tag to use and the packages' initial versions. It allows to control the alpha version bump, for example:
```

```
initial version: 4.61.2

For that initial version, each type of changeset would update as following:
patch: 4.61.3-alpha.0
minor: 4.62.0-alpha.0
major: 5.0.0-alpha.0

If the initial version is already on the alpha tag, either version would bump the alpha.[x]
```

_In our case we want all changes to be major so we work on the following major release._

**Config file explanation**

The main differences on the config file are the following:

```js
{
  "changelog": false // This is set so the changelog of the packages is not updated
  "ignore": [] // We don't want to ignore any package on the next branch
  "fixed": [] // We don't want to have fixed or linked versions of alpha releases
  "baseBranch": "next", // The branch which changesets uses to check versions
  "bumpVersionsWithWorkspaceProtocolOnly": true, // Avoid bumping dependencies on other packages
}
```

> Notes:
> 
> - The changelog is not updated because when we exit the prerelease state the changeset will compile all the changesets and apply onto the new version. eg. 5.0.0 will have all the changesets of the 5.0.0-alpha.\*
> - The `bumpVersionsWithWorkspaceProtocolOnly` makes it possible to avoid bumping packages on the packages that are dependent on the pacakge being updated, this is because we want alpha packages to depend on stable packages, rather than have all of them depending on alpha packages. ie. we would make all packages bump to an alpha core in case we update it.

Reviewer: ```suggestion
> [!NOTE]
> 
> - The changelog is not updated because after exiting the prerelease state, all the changesets will be compiled and applied to the new version. For example, `5.0.0` will have all the changesets of the `5.0.0-alpha.\*`
> - The `bumpVersionsWithWorkspaceProtocolOnly` prevents bumping packages' versions after a package they depend on is updated. This is because we want alpha packages to depend on stable packages. We would bump all packages to an alpha core if we allow it.
```

Trying to make prereleases easier to handle we created a script that you can use on your branch before merging into master, that will bump the package you select, and you can choose if you it's an alpha or beta release, before publishing it to NPM.

You can check the script [here](https://github.com/contentful/forma-36/blob/c6b10071959a085b21e49f5411a5ebff2f8a70d6/scripts/prerelease.mjs)

## Next branch

The next branch is used for the new version of Forma 36, which will probably have breaking changes. To help the development of the new packages, we have setup an automatic publishing of the packages on that branch to the `alpha` tag.
These packages aren't shown on the website as they can have breaking changes often.

### How to automatic publish

> To make changeset publish a new version of the packages it's the same as in the `main` branch, it just requires having a changeset generated and specifing the packages.

Reviewer 2: ❓ will this also be the `npx changeset `command? Maybe would be good to add the exact command then also here in this seciton

Author: Yeah, that is the same command, I'll add the command on the documentation to make it clear

### Exiting prerelease

When exiting the prerelease state, to make the packages stable some steps are necessary:

```sh
# We need to revert the .changeset/config.json file to write the changelog files

npx changeset pre exit # This sets changeset to exit prerelease mode on the next changeset version

npm run changelog:gen # Generates the .changelogrc file that is used to update the What's new page

npx changeset version # This updates the packages changelogs and package.json

npx changeset publish # To publish the packages.
```

_Our CI already handles part of it, we would only need to revert the `.changeset/config.json` file and run the `npx changeset pre exit` command._

If the last commit on the branch before merging into `main` has those changes, the CI will handle the versioning, publishing and generating github releases as usual.

### Differences in the CI

The main difference in the release jobs on the CI, is that in the `next` branch we don't generate the `.changelogrc` file and don't generate the github releases.

Other than that the proccess is the same:

1. Build the packages
2. Update the versioning

> 3. commit the updated package.json file

Author: Let me know if something is not clear, or if you want to change something
